Sources/core/objects/Gun.cpp

```diff
@@ -0,0 +1,96 @@
+#include "Gun.h"
+#include "Tile.h"
+
+Gun::Gun(size_t id) : Item(id)
+{
+    SetSprite("icons/guns.dmi");
+    ammunition_ = 1;
+    max_ammunition_ = 1;
+}
+
+bool Gun::UseAmmo()
+{
+    if (ammunition_ <= 0)
+    {
+        ammunition_ = 0;
+        return false;
+    }
+    --ammunition_;
+    return true;
+}
+
+void Gun::Shoot(VDir target)
+{
+    id_ptr_on<CubeTile> tile = GetOwner()->GetOwner();
```

Comment:
If you think that it is needed to add another param to `ShootImpl` (`ShootImpl(target, type, sound, casing_type)`) then do it. Still, overall there will be finity amount of possible ordinary actions on shot (play sound, create bullet, setup bullet movement, place a casing, ???), and if it will be needed to make some custom action it will be always possible to add it directly to `Shoot` function.